mono/metadata/icall.c

```diff
@@ -9135,52 +9136,80 @@ mono_register_jit_icall_wrapper (MonoJitICallInfo *info, gconstpointer wrapper)
 	mono_icall_unlock ();
 }
 
-MonoJitICallInfo *
-mono_register_jit_icall_full (gconstpointer func, const char *name, MonoMethodSignature *sig, gboolean avoid_wrapper, const char *c_symbol)
+// The few functions that are registered multiple times need to be known here.
+
+void
+mono_no_trampolines (void); // prototype to avoid warning
+
+void
+mono_no_trampolines (void)
 {
-	MonoJitICallInfo *info;
+	g_assert_not_reached ();
+}
+
+// temporary -- later will just be NULL
+static void
+mono_jit_icall_info_free (gpointer info)
+{
+	if (!mono_is_jit_icall_info (info))
+		g_free (info);
+}
 
+MonoJitICallInfo *
+mono_register_jit_icall_info (MonoJitICallInfo *info, gconstpointer func, const char *name, MonoMethodSignature *sig, gboolean avoid_wrapper, const char *c_symbol)
+{
 	g_assert (func);
 	g_assert (name);
 
+	// temporarily allow NULL, until conversion to static storage complete
+	if (info)
+		g_assert (mono_is_jit_icall_info (info));
+	else
+		info = g_new0 (MonoJitICallInfo, 1);
+
 	mono_icall_lock ();
 
 	if (!jit_icall_hash_name) {
-		jit_icall_hash_name = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, g_free);
+		jit_icall_hash_name = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, mono_jit_icall_info_free);
 		jit_icall_hash_addr = g_hash_table_new (NULL, NULL);
 	}
 
-	if ((info = (MonoJitICallInfo *)g_hash_table_lookup (jit_icall_hash_name, name))) {
-		g_warning ("jit icall already defined \"%s\" \"%s\" %p %p\n", name, info->name, func, info->func);
-		g_assert_not_reached ();
+	// Do not allow duplicate registration, either name or function or info,
+	// except the function mono_no_trampolines is reused.
```

Comment:
Which functions are registered multiple times ?

Reply:
There were 3 or 4 but now maybe just 1. There were opcode emulations registered w/ and w/o wrappers, the 2 get_lmf_addr functions, & no_trampoline registered repeatedly.
The 2 lmf functions were probably 2 functions, that I made 1. Compiler/linker could still make them the same even if I resplit them. Gcc won’t do that becuase they are address-taken but msvc still might.
I can report back on the opcode emulation. I hit that back in January when first did this change.

Reply:
https://github.com/mono/mono/pull/12210/files

was the duplicate opcode emulation registration.
I wonder if this is the right/ideal fix — maybe just one name and never wrapper.